PSRabbitMq/ConvertFrom-RabbitMqDelivery.ps1

```diff
@@ -11,26 +11,93 @@
     .PARAMETER Delivery
         RabbitMq Delivery to parse.
 
+    .PARAMETER IncludeEnvelope
+        Include the Message envelope (Metadata) of the message. If ommited, only 
+        the payload (body of the message) is returned
+
     .EXAMPLE
         ConvertFrom-RabbitMqDelivery -Delivery $Delivery
     #>
     param(
-        [RabbitMQ.Client.Events.BasicDeliverEventArgs]$Delivery
+        [RabbitMQ.Client.Events.BasicDeliverEventArgs]$Delivery,
+        [switch]$IncludeEnvelope
     )
     switch($Delivery.BasicProperties.ContentType) {
-        "text/plain" {
-            [Text.Encoding]::UTF8.GetString($Delivery.Body)
+        'text/plain' {
+            $Payload = [Text.Encoding]::UTF8.GetString($Delivery.Body)
         }
-        "application/clixml+xml" {
+        'application/clixml+xml' {
             $XmlBody = [Text.Encoding]::UTF8.GetString($Delivery.Body)
-            [System.Management.Automation.PSSerializer]::DeserializeAsList($XmlBody)
+            try
+            {
+                $deserialized = [System.Management.Automation.PSSerializer]::DeserializeAsList($XmlBody)
+            }
+            catch
+            {
+                #This is for V2 clients...
+                $TempFile = [io.path]::GetTempFileName()
+                try
+                {
+                    $null = New-Item -Name (Split-Path -Leaf $TempFile) -Value $XmlBody -ItemType File -Path (split-path $TempFile -Parent) -Force
+                    $deserialized = Import-Clixml -Path $TempFile
+                    $deserialized = [IO.File]::ReadAllLines($TempFile, [Text.Encoding]::UTF8)
+                }
+                finally
+                {
+                    if( (Test-Path -Path $TempFile) )
+                    {
+                        Remove-Item -Path $TempFile -Force
+                    }
+                }
+            }
+            $Payload = $deserialized
         }
-        "application/json" {
+        'application/json' {
             $JsonBody = [Text.Encoding]::UTF8.GetString($Delivery.Body)
-            ConvertFrom-Json $JsonBody
+            try
+            {
+                $Payload = ConvertFrom-Json $JsonBody
+            }
+            catch
+            {
+                Write-Error 'Invalid JSON. Returning String'
+                $Payload = $JsonBody
+            }
+        }
+        'text/xml' {
+            $Payload = [xml]([Text.Encoding]::UTF8.GetString($Delivery.Body))
         }
         default {
-            [Text.Encoding]::UTF8.GetString($Delivery.Body)
+            $Payload = [Text.Encoding]::UTF8.GetString($Delivery.Body)
+        }
+    }
+
+    if (-not $IncludeEnvelope)
+    {
+        return $Payload
+    }
+    else {
+        return [PSCustomObject][ordered]@{

```

Comment:
The main reason I created this custom object was to reproduce the exact same json (after a convertto-Json) that RabbitMQ output when you create a trace. 
It's very convenient when troubleshooting protocol (which is what I do atm).

Doing what you suggest would create a different structure, which seems to be more or less specific to the .Net client, and the property name would be slightly different (i.e. `ReplyTo` instead of `reply_to`).

Although the `-Raw` (as a different parameter set than `-IncludeEnvelope`) would be a great addition, the `-IncludeEnvelope` is a very useful feature when dealing with foreign protocols (Celery in my case).

Did I convince you? :)